text/0000-io-net-2.1.md

```diff
@@ -0,0 +1,527 @@
+always make sense.
+
+The existing options today will be stabilized as follows:
+
+* `nodelay: bool` - the `TCP_NODELAY` option, and this is only available on
+  `TcpStream`.
+* `keepalive: Option<Duration>` - same as today's `set_keepalive`, but this will
+  differ slightly in terms of implementation with the rest of the socket options
+  here. On Unix `None` will simply set `SO_KEEPALIVE` to `false`, while
+  `Some(dur)` will set `SO_KEEPALIVE` to `true` and `TCP_KEEP{ALIVE,IDLE}` to
+  the specified `dur`. On Windows the implementation will start being wired up
+  to one call to modifying the `SIO_KEEPALIVE_VALS` option (this is
+  unimplemented today). This will be available on `TcpStream`.
+* `read_timeout: Option<Duration>`, `write_timeout: Option<Duration>` - these
+  are covered by [RFC 1047][socket-timeouts] however and already fit within
+  these conventions.  This option will be available on all sockets.
+* `broadcast: bool` - the `SO_BROADCAST` option, only available on `UdpSocket`.
+* `multicast_loop_v4: bool` - the `IP_MULTICAST_LOOP` option, available on
+  `UdpSocket`.
+* `multicast_ttl_v4: u32` - the `IP_MULTICAST_TTL` option, available on
+  `UdpSocket`.
+* `ttl: u32` - the `IP_TTL` option, available on all sockets.
+* `multicast_loop_v6: bool` - the `IPV6_MULTICAST_LOOP` option, available on
+  `UdpSocket`.
+* `multicast_ttl_v6: u32` - the `IPV6_MULTICAST_TTL` option, available on
+  `UdpSocket`.
+* `only_v6: bool` - the `IPV6_V6ONLY` option, available on all sockets.
+
+Some socket options with only setters will be exposed as methods:
+
+* `fn join_multicast_v4(&self, multiaddr: &Ipv4Addr, interface: &Ipv4Addr)` -
+  corresponding to the `IP_ADD_MEMBERSHIP` option.
+* `fn join_multicast_v6(&self, multiaddr: &Ipv6Addr, interface: u32)` -
+  corresponding to the `IPV6_ADD_MEMBERSHIP` option.
+* `fn leave_multicast_v4(&self, multiaddr: &Ipv4Addr, interface: &Ipv4Addr)` -
+  corresponding to the `IP_DROP_MEMBERSHIP` option.
+* `fn leave_multicast_v6(&self, multiaddr: &Ipv6Addr, interface: u32)` -
+  corresponding to the `IPV6_DROP_MEMBERSHIP` option.
+
+[socket-timeouts]: https://github.com/rust-lang/rfcs/pull/1047
+
+### Cross-platform behavior and `IPV6_V6ONLY`
+
+It is pointed out by @stepancheg in [the std::net expansion issue][issue] that
+an IPv6 socket created on Windows will by default not be able to accept IPv4
+connections. This is a cross-platform hazard and may end up causing some code to
+work on Linux (which has the opposite behavior by default) but not elsewhere.
+
+[issue]: https://github.com/rust-lang/rfcs/issues/957
+
+As pointed out [in the man
+page](http://man7.org/linux/man-pages/man7/ipv6.7.html), however, it's possible
+to configure the default at a global level for Linux as well. Despite this, to
+increase cross-platform interoperability, sockets created by `TcpListener::bind`
+and `TcpStream::connect` will have this option set to `false` by default on *all
+platforms*.
```

Comment:
It appears [ruby does this by default](https://github.com/ruby/ruby/blob/trunk/ext/socket/lib/socket.rb#L278-L283) at least, although there's definitely an interesting balance to draw here!